### ISSUE TITLE
HARDWARE RMT ESP32 support...

### ISSUE THREAD

**Issue by sticilface** [Makuna/NeoPixelBus#152]
You probably knew this would come.... :) ESP32 support

any chance?

**Comment by Makuna** [comment]
I haven't received my Esp32 yet; but it is on the list.

**Comment by sticilface** [comment]
:) I just ordered mine from someone in london. not cheap, but same as getting it from china!

**Comment by sticilface** [comment]
so much RAM!

**Comment by AdySan** [comment]
It's been a month, so just checking, I'd really appreciate if this library would work with esp32 arduino port, or better esp-idf itself!

**Comment by Makuna** [comment]
Actively working on this now.

**Comment by AdySan** [comment]
Yay!

**Comment by Makuna** [comment]
https://github.com/Makuna/NeoPixelBus/pull/167
This adds bitbang method as the default, which can support pins below 32.
Hardware support is next

**Comment by Snigelson** [comment]
Do you know yet which pin it will be on?

**Comment by Makuna** [comment]
Hardware support will be able to also work on pins below 32.  There are at least two outstanding issues on the esp32 core iot libraries that need to be addressed before I can publish the hardware support; otherwise the "shared" RMT hardware can only be used by one system at a time (NeoPixel, IR receive, IR transmit, only one at a time, etc).  

Current work is waiting on the following two Esp32 IOF issues.
https://github.com/espressif/esp-idf/issues/514
https://github.com/espressif/esp-idf/issues/517

**Comment by Buffalchill** [comment]
good to hear you are working on it ! Realy would like to see the ESP32 do the Neopixel while receiving data over wifi

**Comment by Makuna** [comment]
@Buffalchill Have you tried the current Esp32 support while doing WiFi?  Theoretically it should work as the Esp32 is dual core.  Its just going to take a lot of the CPU resources.

**Comment by Buffalchill** [comment]
@Makuna since yesterday i am trying out the ESP8266 with your WS2812B driver. It seems to work better than the Neopixel.h  Thumbs up!


When i compile it for the ESP32 i get this ERRROR:

> "In file included from C:\Users\Stefan Schmidt\Documents\Arduino\hardware\espressif\esp32\libraries\LiOr WiFi Projekt\Empfaenger\ESP32_WS2812B_WiFi_17_05_11\ESP32_WS2812B_WiFi_17_05_11.ino:65:0:
> 
> C:\Users\Stefan Schmidt\Documents\Arduino\libraries\NeoPixelBus_by_Makuna\src/NeoPixelBus.h:74:2: error: #error "Platform Currently Not Supported, please add an Issue at Github/Makuna/NeoPixelBus"
> 
>  #error "Platform Currently Not Supported, please add an Issue at Github/Makuna/NeoPixelBus"

**Comment by Makuna** [comment]
@Buffalchill Standard data needed: What device/board are you selecting?  What version of NeoPixelBus library is installed?  Are you using the Arduino IDE or something else?  What version is the IDE?

**Comment by Jorgen-VikingGod** [comment]
I just tried the lib (=master) with my ESP32 devboard and is running fine

**Comment by Buffalchill** [comment]
Hi!  I had an older Lib. now it runs.
But with the ESP8266 my rainbow fader runs well, with the ESP32 it is fading but also blinking pretty strange :-(

**Comment by Jorgen-VikingGod** [comment]
@Buffalchill did you connect it directly on GPIO or through level shifter?
To get stable signals it is always better to serve 5V data line. And of course a 100-220 Ohm resistor is also a good idea as "low-pass" filter to cut the noise.

**Comment by PedroScreamerovsky** [comment]
I'm also having some problems with the ESP32. I'm using a NeoPixel Jewel, and the 4th led is blinking green, with all the example sketches. I've tried with a level shifter but still no luck. Also adding the delay didn't resolved the problem

**Comment by Makuna** [comment]
@PedroScreamerovsky What actual LED model are you using?  WS2813 will require a specific method.

**Comment by Buffalchill** [comment]
I got the same issues with the ESP32 like PedroScreamerovsky.
Now i added a level shifter and a resistor but this didnt realy help. With the ESP8266 the code runs pretty well, but on the ESP32 its flickering on the whole stripe :-/

**Comment by PedroScreamerovsky** [comment]
I'm using these: https://www.adafruit.com/product/2859 (grbw)
I've tried Neo800KbpsMethod, then NeoEsp32BitBang800KbpsMethod (with all the pixels working well except one, that has the green channel doing weird things), and finally the NeoEsp32BitBang400KbpsMethod, that resolves the issue of the green led, but creates lot of random flickering.
As for Buffalchill, the esp8266 is working fine with the same sketch

**Comment by Makuna** [comment]
One more piece of information, how many are in strip?
(the hardware version is pending some improvements from espresif)

**Comment by PedroScreamerovsky** [comment]
It's 7 leds

**Comment by Buffalchill** [comment]
150 leds... pretty much flickering on the stripe

**Comment by sebastius** [comment]
NeoEsp32BitBang800KbpsMethod

NeoEsp32BitBang800KbpsMethod supports any available pin below 32.

Hi, why 'any pin BELOW 32'? Guess who has a board with SK6812 RGBWs on pin 32...

**Comment by Makuna** [comment]
@sebastius Pin 32 and above are handled differently.  The hardware version will use the built in pin matrix switching (allowing hardware to use more pins unlike the esp8266) which also doesn't support 32 and above.

**Comment by Makuna** [comment]
BTW, I just read that the newer WS2812b coming off production are the same as the ws2813, so if the delay after calling show fixes it for you, switch to use the ws2813 method and the delay should not be needed anymore.  They changed the spec so it requires a longer reset time, but didn't change the model number.

**Comment by r-neves** [comment]
I got the same issues with the ESP32 like Buffalchill.
Now i added a level shifter and a resistor but this didnt realy help. With the ESP8266 the code runs pretty well, but on the ESP32 its flickering on the whole stripe :-/

I am using this for the ESP32: 
NeoPixelBus<NeoGrbFeature, NeoEsp32BitBang800KbpsMethod> strip(PixelCount,13);
i am using the WS2812b stripes

ESP32 as some port like ESP8266 that works fine, because using ESP8266 has especific port otherwise strip flickering too.

**Comment by Makuna** [comment]
use this instead...
```
NeoPixelBus<NeoGrbFeature, NeoEsp8266BitBangWs2813Method> strip(PixelCount,13);
```

**Comment by PRO2XY** [comment]
Hi,
I was testing an RMT based WS2812 library for ESP32, which works pretty fine for me (without any level shifter). Maybe it could be of any help to this project?

https://github.com/FozzTexx/ws2812-demo

**Comment by Makuna** [comment]
@PRO2XY If you read the above, this is what is planed.  I have two bugs I need fixed before I can move forward.  Their are two reasons why the RMT that is implemented today are not functional longer term without these.  They are being fixed/applied to the code; but its not ready yet.

1) The RMT format is extremely verbose for NeoPixels.  For strips of any size the memory use is terrible; requiring 4 bytes for every bit of data; or an average pixel would require 96 bytes to represent.  To solve this requires translation on the fly in the RMT ISR.
2) The RMT ISR is not exposed in such a way to be able to share it.  So someone using RMT could only use it for one output, so no IR and NeoPixels together.  That's a waist of the other seven channels of the RMT.

**Comment by r-neves** [comment]
Now I am using a SK6812. With the ESP32 an its flickering on the whole stripe :-/

I am using this for the ESP32:
NeoPixelBus<NeoGrbwFeature, NeoEsp32BitBang800KbpsMethod> strip(PixelCount,13);
i am using the WS2812b stripes

ESP32 as some port like ESP8266 that works fine, because using ESP8266 has especific port otherwise strip flickering too.

How can fix it and what is the correct pin to use?

**Comment by Makuna** [comment]
@r-neves I believe I gave you something to try above, but you may have missed it.
```
NeoPixelBus<NeoGrbFeature, NeoEsp8266BitBangWs2813Method> strip(PixelCount,13);
```
Some of the new chips are using the newer Ws2813 protocol.

**Comment by r-neves** [comment]
But my strip is SK6812, and stills flicking.
If I turn Red and Blue flicker, but Green and White no flicker

**Comment by Makuna** [comment]
@r-neves Are you stating that the simple test (without WiFi) the problem also exists?  If so, please join on the gitter channel to diagnose your issue, as its not the same as what this issue is tracking.

**Comment by r-neves** [comment]
My code is very simple
`#include <NeoPixelBus.h>

const uint16_t PixelCount = 144; // make sure to set this to the number of pixels in your strip
NeoPixelBus<NeoGrbwFeature, NeoEsp32BitBangWs2813Method> strip(PixelCount, 14);

void setup()
{
  strip.Begin();
}

void loop()
{
  for (uint16_t i = 0; i < strip.PixelCount(); ++i)
  {
    RgbwColor color = RgbwColor(0, 255, 0, 0);
    strip.SetPixelColor(i, color);
    strip.Show();
  }
}`

**Comment by Makuna** [comment]
@r-neves This "issue" had changed to tracking hardware support, I am sorry this was not obvious and I changed the title to make it so.

Please do as I ask and join the gitter channel for support **for your specific problem**; as often there are problems with connectivity and powering the NeoPixels that cause most of these flickering issues.

**Comment by phenzao** [comment]
I am using the same example like r-neves above with esp32 dev board and platformio and I get this error: "undefined reference to `bitbang_send_pixels_800'"

**Comment by EBZ-Krisemendt** [comment]
stumbled across this issue while working on a similar problem with faulty pixels on esp32.
as a workaround (while hardware-support isnt ready) adding the delay(1) right before strip.Show() as descripted in issue https://github.com/adafruit/Adafruit_NeoPixel/issues/139 help me to resolve the problem and all pixels are showing correctly.

**Comment by Makuna** [comment]
Thanks for the reference.  I was afraid of something like this coming along, and with this knowledge I will make the hardware support the default in the future. 

I really wish they would fix the RMT support so I can finish the hardware version and get it out for people to use.

**Comment by EBZ-Krisemendt** [comment]
is it to soon to ask for any updates already since the related issue https://github.com/espressif/esp-idf/issues/514 got closed? ;)

**Comment by Makuna** [comment]
I still need [espressif/esp-idf#517](https://github.com/espressif/esp-idf/issues/517) fixed.

**Comment by Makuna** [comment]
Related issue https://github.com/Makuna/NeoPixelBus/issues/182 to this conversation; I will be updating the bitbang with a fix soon while I am still waiting for a fix from the SDK as linked above for the hardware support.

**Comment by chrisella** [comment]
Hey,
Any news on this?
I have a 240 pixel strip which works almost fine using a combination of the delay(1) and NeoPixelBus<NeoGrbFeature, NeoEsp32BitBangWs2813Method> strip(PixelCount, PixelPin);

The only problem is 1-2 pixels don't behave and are permanently turned on, usually earlier in the chain in-case that has any bearing. This is using the ESP32 without any kind of WIFI features, as soon as I enabled that I began having all manner of problems again with the whole strip flickering etc.

**Comment by Makuna** [comment]
I still need espressif/esp-idf#517 fixed.
Please go to the above issue and up vote it.  It has been pretty stagnate for awhile.

**Comment by chrisella** [comment]
Thanks, will do, thanks for continuing to push the ball on this!

**Comment by vifino** [comment]
With the git version and `NeoPixelBus<NeoGrbFeature, NeoEsp32BitBang800KbpsMethod> strip(LEDS, PIN);` my 256 led ws2812b matrix has minimal errors (ie pixel set to weird/random value) and no flickering.
Happens less often than once a second when updating everything at ~60hz.

Before, I didn't have the git version and didn't select the bit banging method. It was unusable: Very heavy flicker and lots of pixel corruption.

**Comment by mcer12** [comment]
Seems like things started moving on espressif side. @Makuna

**Comment by Makuna** [comment]
Yes!

**Comment by alexkirillov** [comment]
ESP32-digital-RGB-LED-drivers, mentioned earlier by PRO2XY is the only (arduino, didn't try any esp-idf) library that I was able to make work with my ESP32 and 3x 241 pixel LED rings. 

All other libraries, unfortunatly including NeoPixelBus at the moment, give arbitrary timing errors.
Althru they are rare, like 1 wrong pixel in 1000.
But given the setup is 3x 241, it occures every second time I'm setting lights to rgbw(0,0,0,0)

i understand that @Makuna is waiting for @igrr to optimize RMT stuff, not to waste all the memory on pixel data.

**Comment by JanVelleman** [comment]
Hi Makuna,
Any news about this?

Sincerely

**Comment by iontichy123** [comment]
Also interested in this topic. Any plans?

**Comment by Makuna** [comment]
I still need espressif/esp-idf#517 fixed.

**Comment by r1dd1ck** [comment]
Just to let everyone know - the required changes have been recently merged with the IDF master, so it is now only on the folks maintaining the **`arduino-esp32`** package to update it. 

Could be days, but could also be months - hard to tell :-s But I think they are currently already behind schedule with IDF updates, so maybe it will be sooner than later :-)

PS: I have the changes already incorporated into NeoPixelBus and currently running 8-way parallel without breaking a sweat, so good news indeed :-)

**Comment by Makuna** [comment]
I also have a i2s implementation ready (based on the esp8266 DMA) but the esp32 team has asked me to hold off publishing it so that they can wrap the guts of it into an API.

**Comment by MichMich** [comment]
Hi Guys! Any status update on this? I'm really in the need of a fix. Thanks!

**Comment by r1dd1ck** [comment]
@MichMich 
The **RMT** implementation will need to wait till [esp-idf@5bc8ae3](https://github.com/espressif/esp-idf/commit/5bc8ae3b3a8be3e276c3318865efa9f0b5d82720) hits the `arduino-esp32` core. You can go to the [arduino-esp32](https://github.com/espressif/arduino-esp32/) repo and bugger them about it - which might speed things up a bit (or not) :-D

The **I2S** implementation mentioned by @Makuna, on other hand, would probably not require any changes to the current core, so it should be ready to go. But only in case the author is nice enough to share a pre-release version, of course ;-)

**Comment by MichMich** [comment]
@Makuna, is it possible for you to share the i2s solution?

**Comment by Makuna** [comment]
@MichMich The good news is that I have been given the go ahead on the i2s implementation; so it should show up within a few days.

**Comment by MichMich** [comment]
@Makuna Awesome! Looking forward to it! :)

**Comment by Makuna** [comment]
NEW UPDATE COMING:  I just merged in Esp32 Hardware Support using I2S (very similar to how Esp8266 was done but without the pin restrictions).  You can access it today by using Github to get the latest code.  A new library (through the Arduino IDE) will be made later today and might take a day or two to show up.

The new default on Esp32 will be this I2s method, using the i2s bus 1.

This ISSUE will remain open tracking RMT hardware support.

**Comment by MichMich** [comment]
@Makuna I can confirm that your I2S implementation solves the issue! Awesome work! Thanks!

**Comment by PeterOFre** [comment]
Hi all, I am new in GitHub and new in ESP32.  Have created a Light Painting Stick based on an Arduino Uno and the Adafruit Data Logging Shield  following the Adafruit Tutorial and it is working fine except the conflict with the I2C LCD Display. Have thought I am using ESP32 and bought a M5Stack Module with inbuild SD card, Display and 3 buttons. The NeoPixelBus demo is working fine on the M5Stack, but again, the LCD Display and SD Card reader stop working when the  strip.Begin() and strip.Show(); command has reached. Is there a way to have NeoPixel and LCD + SD Card working in parallel ? I could upload the Arduino sketch if needed.

**Comment by Makuna** [comment]
@peterofre please use gitter (see readme, wiki) to ask questions that are not specific to this issue.

**Comment by keavin7** [comment]
@Makuna I'm having an issue with I2S on ESP32. It works fine in the test standalone script, however when I included it in my own project it crashes immediately. Including the stack trace below.

```
Guru Meditation Error: Core  1 panic'ed (LoadProhibited)
. Exception was unhandled.

0x400d4748: i2sWriteDone at C:\Users\Sup\Dropbox\workspace\Beem\Beem\code\Beem_V1/src\internal/Esp32_i2s.c line 380
0x400d4748: i2sWriteDone at C:\Users\Sup\Dropbox\workspace\Beem\Beem\code\Beem_V1/src\internal/Esp32_i2s.c line 380
0x400d4d45: NeoEsp32I2sMethodBase ::IsReadyToUpdate() const at C:\Users\Sup\Dropbox\workspace\Beem\Beem\code\Beem_V1/src/internal/NeoEsp32I2sMethod.h line 115
:  (inlined by) NeoEsp32I2sMethodBase ::Update() at C:\Users\Sup\Dropbox\workspace\Beem\Beem\code\Beem_V1/src/internal/NeoEsp32I2sMethod.h line 127
0x400d4dba: NeoPixelBus   >::Show() at C:\Users\Sup\Dropbox\workspace\Beem\Beem\code\Beem_V1/src/NeoPixelBus.h line 147
:  (inlined by) Lights::show() at C:\Users\Sup\Dropbox\workspace\Beem\Beem\code\Beem_V1/src/lights.cpp line 26
0x400d5441: ModeManager::render() at C:\Users\Sup\Dropbox\workspace\Beem\Beem\code\Beem_V1/src/modes.cpp line 271
0x400d3171: Beem::update() at C:\Users\Sup\Dropbox\workspace\Beem\Beem\code\Beem_V1/src/beem.cpp line 71
0x400d56ae: loop() at c:\users\sup\.platformio\packages\toolchain-xtensa32\xtensa-esp32-elf\include\c++\5.2.0\ext/new_allocator.h line 110
0x40166a46: loopTask(void*) at C:\Users\Sup\.platformio\packages\framework-arduinoespressif32\cores\esp32/main.cpp line 18 (discriminator 1)
```

My guess is there's some kind of conflicting hardware configuration, but I just don't know anything about I2S so i'm not sure where to look around, and the lack of detail in the trace doesn't give me alot to go on either. Neither of the bus's work.

While the error mentions, i2sWriteDone, line 380 seems like a more likely place for a failure.

'''
void i2sInit(uint8_t bus_num, ...) { ...
    esp_intr_enable(I2S[bus_num].isr_handle);
}

I believe the only other user set interrupt i have is on pin 0 (sparkfun button), however this same functionality works in a modified test file. The other culpret i think it might be is wifi... I have a few popular web services included, I doubt these would drive this issue.

AsyncWebServer
AsyncEventSource
AsyncWebSocket
SimpleBLE

Wondering if you might have an idea about whats going on? I'll try to address the com libraries by myself

Thanks :)

**Comment by keavin7** [comment]
Removing the libs didn't help... hmmm

**Comment by keavin7** [comment]
Does the template need to be globally declared in the main file?

**Comment by keavin7** [comment]
@Makuna Solved this problem, after chugging a gallon of coffee I realized like a fool I didn't call 'strip.Begin()'

**Comment by keavin7** [comment]
@Makuna Been testing this code out. Using the default <NeoEsp32I2s1800KbpsMethod> i get flickering on the lights when serial is sending. Is anyone else experiencing something like this?

**Comment by Makuna** [comment]
I have not heard of any issues.  Can you provide a minimum sketch that reproduces the problem?  With the minimum sized strip if that also is a culprit.

**Comment by Xonotron** [comment]
While evaluating different libraries for a project, I stumbled upon an issue with ESP32, SK6812 LEDs and NeoPixelBus.

I reduced my project to the most simple code that reproduces the problem: strip.Show() seems to be ignored every now and then (1 out of 10 roughly) unless I call strip.Begin() every time prior to that. NeoEsp32BitBang800KbpsMethod works as expected.

Here's the code:
```
#include <NeoPixelBus.h>

NeoPixelBus<NeoGrbFeature, Neo800KbpsMethod> strip(1, 4);

int brightness=32;
RgbColor yellow(brightness, brightness/4*3, 0);
RgbColor blue(0, 0, brightness);

portMUX_TYPE mux = portMUX_INITIALIZER_UNLOCKED;
unsigned long interruptTime=millis();
bool blinkNow=false;
unsigned int resetTimer=millis();
bool resetLed=false;


void IRAM_ATTR handleInterrupt() {
  portENTER_CRITICAL_ISR(&mux);
  if(millis()>interruptTime+200) {
    interruptTime=millis();
    blinkNow=true;
  }
  portEXIT_CRITICAL_ISR(&mux);
}

void setup() {
  pinMode(12, INPUT_PULLUP);
  attachInterrupt(digitalPinToInterrupt(12), handleInterrupt, FALLING);

  strip.Begin();
  strip.Show();
}

void loop() {
  if(blinkNow) {
    portENTER_CRITICAL(&mux);
    blinkNow=false;
    portEXIT_CRITICAL(&mux);
    
    resetLed=true;
    resetTimer=millis();
    //strip.Begin();
    strip.SetPixelColor(0, yellow);
    strip.Show();
  }
  if(resetLed) {
    if(resetTimer+100<millis()) {
      //strip.Begin();
      strip.SetPixelColor(0, blue);
      strip.Show();
      resetLed=false;
    }
  }
}
```

...and the hardware setup:
![esp32_setup_](https://user-images.githubusercontent.com/42165943/44308854-68f78b00-a3bd-11e8-9bc3-3e9ec40155ff.jpg)

@me-no-dev said that NeoPixelBus uses an early version of the I2s driver (https://github.com/adafruit/Adafruit_NeoPixel/issues/139#issuecomment-403378252), could this be an issue there?

**Comment by Makuna** [comment]
@Xonotron This issue is about having support for Esp32 at all, not really a general issue to ask questions about something not working correctly.  There are a lot of pitfalls with using NeoPixels and if you glance at the Wiki  FAQ, you can see solutions to most of them.  The gitter channel is the best place to start in asking for help and then creating an issue once it has been identified to be a real issue.  The i2s version I have works great and this issue is tracking RMT version I plan to include.

**Comment by Makuna** [comment]
@r1dd1ck implementation for what?  RMT?  Already have [Pull 202](https://github.com/Makuna/NeoPixelBus/pull/202) but I am waiting for latest support to be issues in an official Esp32 Arduino release.

**Comment by Makuna** [comment]
@r1dd1ck That's great!  Go ahead and create a pull.  While I did start one (that lead to idf requests) I haven't updated it yet and I suspect it will require some significant changes.

**Comment by r1dd1ck** [comment]
@me-no-dev 
can you please explain what exactly you mean by "clash with the Arduino API"? 

Isn't the upcomming Arduino RMT "driver" supposed to be just a wrapper for the IDF stuff, with the sole purpose to make it "easier to use" for the average arduino user? Or are there some aspects to it which will render the use of the native API a "high-risk" endeavour? 

I would certainly hope that's not the case, and that the final version will be sanitized enough to avoid any such conflicts..

**Comment by me-no-dev** [comment]
when you control the same resource from two different places, you can surely end up in position where your stuff does not work. I personally can go just fine with a mix of IDF, Arduino and my own code, but many users will have issues if peripherals are used by libs and hide away the choices that they made for pins, channels and so on. The Arduino RMT driver actually keeps track of the used channels and memory and will not allow two libs to try to use the same channel and will in fact give then different resorces so both can function. Arduino and IDF users are different in skills and more.

**Comment by r1dd1ck** [comment]
@me-no-dev 

> hide away the choices that they made for pins, channels and so on

• there is no sane case of "hiding away of pin choice" I could possibly think of.. and
• channel conflicts can be safely sanitized (I know because I'm already doing it in my code)

So to me it seems that the real problem here is not whether `my code` is going to clash with the `Arduino API` (because my code is already sufficiently sanitized for this), but rather whether the `Arduino API` is going to clash with `my code` ... Splendid :-D

I hope most of these possible "clashes" will be resolved in the `Arduino RMT API` before it gets finally pushed to `master` .

But I think the bottom line here is - that even if the `Arduino API` is not sanitized to avoid conflict with the native API, nothing is stopping us from updating our code to automatically "register" the used RMT resources within the `Arduino RMT API`, once it becomes available..

@Makuna 
What's your take on this? 
Want to wait for the Arduino implementation (with all the added overhead)? 
or would it be OK to use the native approach - which will later get adjusted to avoid possible clashes with the Arduino API once it becomes available (in case the Arduino API does not implement these checks) ?

Let me know :-)

**Comment by r1dd1ck** [comment]
@me-no-dev 
> I strongly believe that the RMT Arduino lib is much better and easier to use by Arduino users.

and I did not question that aspect at all. I, too, believe it will.

> I am not really sure about what overhead you are talking about.

all the overhead a wrapper usually takes. Unless it is not a wrapper, and you are re-writing the whole driver from scratch, doing it in a more efficient way than the native API (unlikely).

> You do not need to care what other libs are using RMT, how much memory they have dedicated to their channel and so on.

Oh I believe you do have to care, because all the currently available libs already use the `native approach`, and you can't simply expect all of the maintainers to promptly rewrite their code just to avoid "clashes" with some newly introduced API's. The point is, that if the new API is not properly sanitized, it will make all those libraries which already make use of the `RMT peripheral` potentially unstable - and I think this is a scenario the `Arduino-ESP32` team should be trying to avoid.

Don't understand me wrong, I fully support the idea to offer easier ways for Arduino users to utilize the potential of all the available peripherals - but I also think it should be done in a "transparent" way that does not necessarily introduce conflicts with the native API.

**Comment by r1dd1ck** [comment]
@me-no-dev 
to reduce the whole matter down to the most important point:

> Make sure the new Arduino API's don't "clash" with resources allocated by the native approach (IDF), and all will be fine.

:-)

**Comment by Makuna** [comment]
My take on this:
I spotted the power of RMT, but also the conflict and pain when two library writers use it, so I avoided creating a bigger problem down the road by seeking improvement (the issue to fix it) rather than jumping on it to use.  These problems are what they are fixing and I support that.  

Further, the Bitbang was working well enough; until it didn't.  But I now have i2s hardware support so there is no strong need to force conflict and have RMT support.  

As a library writer, its my problem if I work around standard APIs (or non-existent).  While its fun to hack something together and I don't blame people for doing so, I understand the risk of it becoming incompatible and it will require me to "fix" it.  Other library writers should understand this also.  So I believe a low priority should be on the API creators to keep it compatible with my workarounds (within the context of Arduino support) as long as they are solving the issues of sharing and functionality.  I don't want them creating "crap" APIs due to some library writer doing something completely wrong.

SO, I will hold off support for it until the new stuff arrives.

**Comment by r1dd1ck** [comment]
@Makuna 
I don't think I would call usage of the native API a "workaround", or even "doing something completely wrong" - as you state. However I understand your concerns about including something that could become broken by some future changes to the Arduino-ESP32 package (as hinted by @me-no-dev earlier), so fair enough.

> These problems are what they are fixing and I support that.

Yes, although I'm afraid those "problems" will probably remain as there is alot of libraries out there succesfully using RMT natively already. ESP32 might be obsolete by the time most of them do the switch (if ever) :-/

> Further, the Bitbang was working well enough; until it didn't.

I would say that strongly depends on your definition of "well enough" :-D 

> But I now have i2s hardware support so there is no strong need to force conflict and have RMT support.

Can't really compare the max. 2 channels of `I2S` with the seamless 8 of `RMT`. A couple hundred pixels is simply not enough for any larger projects. I wish the china-men would release a higher bitrate version of the pixels already, with at least 12bit+ resolution :-/

**Comment by Makuna** [comment]
@r1dd1ck I created a new branch, labeled [Esp32RMTExperimental](https://github.com/Makuna/NeoPixelBus/tree/Esp32RMTExperimental), that you can create a pull against with your RMT code.  This will allow advanced users to play with it until we get the offical Arduino support.  This will limit exposure from the novice users and thus save me some support effort.

**Comment by r1dd1ck** [comment]
Just had a look at the proposed `Arduino RMT driver` and..

It seems they are completely rewriting the API from scratch into a new Arduino incarnation, which for the most part essentially just mirrors the native API including data structures (just under different names) - so any added value in this sole regard is escaping me. 

There are of course benefits, like the already mentioned resource management, introduction of simpler methods for init, and to a certain extent also send & receive. So that's a plus.

But I am sad to inform that, at least in its current incarnation, it will be barely usable for any meaningful NeoPixel use :-/ Not going much into detail here as the reasons should be more than obvious after a quick glance at the source. I will only hint that it is "back to square # 1", rewinding to a state the native API was roughly 1.5 years ago..

Holding off judgement until the final version appears though. Fingers crossed :-)

**Comment by me-no-dev** [comment]
yet one of the examples in that PR is exactly neo pixels and on top of that the lib was written to be able to stream large amounts of pixel data (which the IDF implementation does not really do well). In any case... neo pixels are working for me with both i2s and rmt ;)

**Comment by r1dd1ck** [comment]
> yet one of the examples in that PR is exactly neo pixels and on top of that the lib was written to be able to stream large amounts of pixel data (which the IDF implementation does not really do well). 

OK. Funny you mention the example from PR, because that was exactly the trigger which prompted me to investigate further on it.

I just hope that @Makuna will have more success in spotting what your magnifficient take on re-inventing the wheel is missing, cause I surely ain't gonna spill the beans now :-)

**Comment by Makuna** [comment]
i2s hardware support has been present for awhile now.  Updated title to track adding RMT support

**Comment by Makuna** [comment]
https://github.com/Makuna/NeoPixelBus/pull/272

**Comment by Makuna** [comment]
Just a link before I close this as the solution is now live.  Just a little problem noted in RMT apis cause some non-optimum perfromance, but it still works with all 8 channels.

https://github.com/espressif/arduino-esp32/issues/2885
